modules/imap/functions.php

```diff
@@ -1321,6 +1321,49 @@ function snooze_message($imap, $msg_id, $folder, $snooze_tag) {
     }
     return $res;
 }}
+if (!hm_exists('add_label_to_message')) {
+function add_label_to_message($imap, $msg_id, $folder, $tag) {
+    if (!$imap->select_mailbox($folder)) {
+        return false;
+    }
+
+    $msg = $imap->get_message_content($msg_id, 0);
+    preg_match("/^X-Cypht-Labels:(.+)\r?\n/i", $msg, $matches);
+
+    if (count($matches)) {
+        $msg = str_replace($matches[0], '', $msg);
+        $tags = explode(',', $matches[1]);
+        if(in_array($tag, $tags)) {
+            unset($tags[array_search($tag, $tags)]);
+        }else{
+            $tags[] = $tag;
+        }
+    }else {
+        $tags = array($tag);
+    }
+    $msg = "X-Cypht-Labels:".implode(',', $tags)."\n".$msg;
+
+    $msg = str_replace("\r\n", "\n", $msg);
+    $msg = str_replace("\n", "\r\n", $msg);
+    $msg = rtrim($msg)."\r\n";
+
+    $res = false;
+
+    if (!count($imap->get_mailbox_status($folder))) {
+        $imap->create_mailbox($folder);
+    }
+    if ($imap->select_mailbox($folder) && $imap->append_start($folder, strlen($msg))) {
+        $imap->append_feed($msg."\r\n");
+        if ($imap->append_end()) {
+            if ($imap->select_mailbox($folder) && $imap->message_action('DELETE', array($msg_id))) {
```

Comment:
I think you should only select the mailbox once at the top of the method. No need to further select the same folder as you are working in one folder only.

Reply:
Removed